### DISCUSSION TITLE
Branching Strategies

### DISCUSSION THREAD

**Discussion by Elliria** [autokey/autokey#823] [Maintenance goals]
I've been trying to learn what I can about branching strategies in an effort to understand how our branches work. As part of that, I visited the **GitHub** channel on IRC. They recommended that we switch to what I later found out is known as the **feature branching** strategy and it sounds like it would be a very good idea. A bit of poking around online to get more information on how that works turned up this breathtakingly-awesome video on branching strategies from a man named Victor: [https://www.youtube.com/watch?v=U_IFGpJDbeU](https://www.youtube.com/watch?v=U_IFGpJDbeU)

These are the six strategies described by Victor, listed in frustration order, from the **least frustrating (1)** to the **most frustrating (6)** and you'll see where our project falls in the list:

1. In **trunk-based** development, there's a main branch and everybody directly makes commits to the main branch without the need for pull requests, merges, etc. This is a low-stress environment.
2. In **feature branching** (AKA **GitHub flow**), there's a main branch and when anyone wants to make a change, a feature branch is created, worked on, merged into the main branch with a pull request, and then deleted. This is a low-stress environment.
3. In the **forking** strategy, there's a main branch and when anyone wants to make a change, the code gets forked, worked on, and merged into the main branch with a pull request. This is a low-stress environment.
4. In **release branching**, there's a main branch and a branch for each release and branches for hotfixes, with each branch being worked on and merged into the main branch with a pull request. The work between teams or developers is disconnected and the branches can become conflicted, delaying integration and causing confusion. This can be useful for teams that create and support multiple releases and apply hotfixes to them. This is a medium-stress environment.
5. In **Git flow**, there's a main branch and a development branch. Feature branches and release branches get created off of the development branch. Release branches get merged with the development branch and with the main branch. The work between teams or developers is disconnected and the branches can become conflicted, delaying integration and causing confusion. We should think long and hard about whether it's worth spending time with such a branching strategy unless we're there to help the project get rid of it. This is a high-stress environment.
6. In **environment branching**, there's a release branch, a development branch off of that, a branch for every environment off of that, and hotfix branches off of those, and everything gets merged everywhere and releases are being developed in parallel. It's absolute madness. This is a high-stress environment.

Victor's recommendation for branching strategies:
* You should be using **trunk-based branching** strategy if you don't care about quality... or... if you're a superhero and know your way confidently around in the code, consistently do high-quality work, trust your testing process, etc. Most of us don't fall into this category. 
* You should be using **feature branching** or **forking strategy** if you're working in small, self-sufficient teams with potentially small applications and you know how to split working into small pieces and, perhaps, sometimes need to apply hotfixes for backwards-compatibility.
* If your project doesn't fit into either of those two choices, you should either choose one of them or run away and don't look back. According to Victor, life is too short to spend on merging branches and resolving conflicts.

Alas, we use number **5** - the **Git flow** branching strategy, which is one of the two most frustrating strategies on that list. Not only that, but we don't quite use it as described. We have three long-running branches (master, develop, and beta) and a few feature branches. Unlike the standard structure for this branching strategy, although our **develop**  branch is based on the **master** branch, our **beta** branch is **also** based on the **master** branch instead of being based on the **develop** branch as it should be, which makes them independent of one another and causes them to go in completely different directions. We then use the **feature-branch** or **forking** strategy to make changes to each of those long-running branches rather than committing to them directly. We also have what's been referred to as **Continuous Integration (CI)**, but which is actually **Continuous Deployment (CD)**, provided to our project by our **beta** branch that makes it so that tagged releases that are merged into develop, beta, or master will automatically be built.

> **Note:** In the event that you enjoyed the above video, Victor has another video that's equally as excellent on **Continuous Integration vs Continuous Delivery vs Continuous Deployment vs Testing** and that I highly recommend: https://www.youtube.com/watch?v=0ivcSjpUzl4

The folks on IRC feel that we should be using the **feature branching** strategy, which is exactly what Victor recommended above. The IRC folks also feel that our **CI/CD** should be provided by the **master** branch instead.

I'm thinking that both Victor and the folks on IRC are making a lot of sense. I'd love some feedback on it all.

**Comment by josephj11** [comment]
Initial thoughts. I haven't watched the videos yet.

My main concern is release stability. The first three approaches don't even have a develop branch. That means that when we release something (create debs, etc.), main effectively gets frozen in those debs, etc. That sounds fine, but AFAIK, packagers don't use our debs, etc. they make their own whenever they get around to it.

IIUC, this has a few implications:

If we continue development, then packagers get whatever they get depending on when they get it, not on when we release it. This might mean that 0.96.1 on Debian might be dfferent than 0.96.1 on Fedora... That way leads to madness, especially for people like me trying to do support, but also for developers trying to figure out what a line number in a traceback from an issue report actually refers to in the version of the code they're looking at.

Currently, I'm not sure how it works, but I think that things can be added to develop whenever someone thinks they're good enough to share with the rest of the team. This does not imply that they are good enough/ready/agreed upon to be included in the next release. That can be decided later when we intentionally create a release.

If we go with any of the first few options presented above, we lose the buffer that develop provides and we lose the ability to add new features and fixes that are still tentative or maybe for a release that's further out.

These concerns need to be addressed by whatever strategy we adopt.

And there's the significant issue that any changes to our build/release strategies have to be coded/configured by someone. Who, besides @BlueDrink9 (who probably doesn't want to do that work) knows how to do that?  Maybe @sebastiansam55 does. @Elliria  is learning fast, but I'm not sure she's ready for this yet.

**Comment by Elliria** [comment]
Thank you for the initial feedback. I'd love to hear more after you've watched the video(s). Releases shouldn't be a problem in any of those scenarios, because they're created with tags in all branching strategies regardless of how many branches exist.

My main concern is that the branches are in disarray and it's likely to only get worse. I've contributed to some of that disarray and am likely to continue to do so when we need to make changes to more than one branch. When I've turned to others for advice on how to do that in a tidy way, I keep getting the same answer and that's to simplify things - most specifically, the branch strategy. They don't actually tell me how to clean up or "properly" use what's there because they don't feel that we should be using it.

My hope is to reach a point where it's only necessary to contribute content to one branch (the **develop** branch), but that's not likely to be possible with this branching strategy. There will inevitably be times when the same content is also needed in one or more other branches, as we've been seeing with some of the updates we've been doing.

When it comes to making the same change in more than one of (or all three of) our long-running branches, would you like me to continue to add to the branch divergence by copying known-good contents from a file in one branch to the same file in another branch in separate commits and pull requests? Or do you have a different approach you'd like me to use?

**Reply by josephj11** [reply]
I'll watch the videos when I get a chance. Hopefully some or all of my concerns are non-issues. Since these concerns are almost completely generic, I'm sure they can be addressed.

Given my lack of experience with git and GitHub, I'm not asking you to do anything. I am concerned that unnessary complexity is being added by the way we're doing things now and I want to make sure any changes we make actually make things better without creating new problems that might be worse.

I do not consider beta as a valid branch except when a new release is immanent, so I don't want anything done to it when it's not a proposed release under final testing. That leaves us with two branches, main and develop. Apparently, we can't delete beta when it's not in use (which is what I wanted to do).

**Comment by Elliria** [comment]
No hurry. Just beware that things will get messier before they get tidier. Although you don't consider **beta** to be a valid branch, it has been basically forced upon us by the structure of our branching strategy, so it's a necessary evil. As such, there are things that will have to happen to it, because the fact that it needs to be there means it gets used. It's currently behind on the action versions that I just updated in the **develop** branch and will need to have that done to it as well. This is because those actions will also be run on the **beta** branch. Ditto for the **master** branch.

So, I'm back to my original question: When a change (like a version update) needs to be made to the same file in more than one branch, would you like me to continue to add to the branch divergence by doing the appropriate edits from the list below in separate commits and pull requests?
* Make **identical** changes to the same file in the various branches **if they're identical to begin with.**
* Make **custom** changes to the same file in the various branches **if they're not identical to begin with**.

**Reply by josephj11** [reply]
I don't have an opinion on this. The more multibranch stuff we do, the worse unraveling it all becomes when we need to synchronize everything, but if we don't do it, the things might slip through the cracks. I'd like some input from @BlueDrink9 and @sebastiansam55.

This really depends on how it all affects our workflow and QA and since I don't do much of that, I don't know what I think would be best.

**Reply by sebastiansam55** [reply]
I err on the side of less git stuff, because I don't have much experience with it. At most I would have a master/develop branch and even then, what % of users are installing autokey directly from master branch? 

I would assume 99% would be installing from their package manager of choice so at that level there isn't even much reason to keep master "stable/clean" if we use git tags to tag releases. 

What's the current status of the beta branch? They were beta of the .96 release it seems like? So they could be deleted since .96 has been released?

**Reply by Elliria** [reply]
> I don't have an opinion on this.

What will happen if I go for it and submit pull requests to, for example, do the same update to the activities in the **beta** and **master** branches as just got done to the **develop** branch? Will those be cancelled, just sit there, or get approved? Or would you rather I hold up on stuff like that until further notice after you get more input from others?

**Reply by Elliria** [reply]
> I err on the side of less git stuff, because I don't have much experience with it. At most I would have a master/develop branch

Thank you for that. I agree, and I think @josephj11 does, too. Although, then again... I think even that's too much and am really liking the **feature branching** strategy above as the gentlest one.

> even then, what % of users are installing autokey directly from master branch?

No idea, but you're probably right that most of them are grabbing the AutoKey packages provided by their operating systems.
 
> What's the current status of the beta branch? They were beta of the .96 release it seems like? So they could be deleted since .96 has been released?

@BlueDrink9 [said we shouldn't delete **beta** because workflows make use of it](https://github.com/autokey/autokey/pull/811#issuecomment-1472725923). Also, if you look at [our branches](https://github.com/autokey/autokey/branches) both **develop** and **beta** are based off of **master** and are independent of one another. As a result, they're each ahead of and behind **master** in differing ways and each seems to have things the other doesn't. Some of what each of them has (and no longer has) may be "good stuff" and not something to be thrown away, but someone will have to look it all over and figure that out one file at a time. It will not be a pretty job.

I would have thought that **beta** shouldn't have anything in it that isn't in **develop**, but **develop** would be expected to have some things that aren't in **beta**.

The way I envision **Git flow branching** working, if our three long-running branches must continue to exist, would be for **beta** to be based off of **master** and for **develop** to be based off of **beta**. All development would only ever happen to **develop** from pull requests from temporary branches that get deleted afterwards. Only production-worthy changes in **develop** would go onto **beta** via pull requests. That way, **beta** wouldn't contain anything that's not already in **develop**, **develop** would contain everything that's in **beta** and also things that either aren't up to snuff yet or at all and/or aren't finished. That would make comparisons of the two branches much gentler.

I'm still unclear as to the purpose of **master** or how it would fit into this scenario in any useful way, so I'll just figure that it gets to sit there looking pretty.

**Reply by josephj11** [reply]
> What's the current status of the beta branch? They were beta of the .96 release it seems like? So they could be deleted since .96 has been released?

I wanted to delete the beta branch as well until @BlueDrink9 told us not to. Unless we use it as "next" release (not specifically a beta), I would leave it untouched until we have an actual need for a beta of the next release and, ideally, I would "copy" all of that new release onto an "empty" beta so there is no question of syncing anything.

If we go with @Elliria 's suggestion of develop->next->main that would require that fixes to the beta or main would have to be applied first in develop (if possible), then migrated to next and finally, to main itself. That would keep all PRs one way with the required discipline, but that sounds like a lot of work.

**Reply by Elliria** [reply]
I'm not actually suggesting or recommending **master** <-- **beta** <-- **develop**, but am seeing it as the least-evil way of restructuring the **master** <--> **beta** <-->  **develop** <--> **master** that we currently have), since it would then mean that data travels in only one direction, making it easy to know where a change needs to happen and making comparisons between branches much easier.

As for what I'd recommend, I'm convinced that **feature branching** is the way to go and that **master** <-- **temporary feature branches** is the gentlest structure.

**Comment by josephj11** [comment]
Forgive me for not having had time to watch those videos yet.

I'm not concerned (much) about users installing from the master branch directly. I am concerned that distro packagers will build/install from main. If we use release tags (which I don't know anything about yet), will a packager accessing main reliably get the version of the code that we intended to release or will they potentially get changes that were made after the the release was tagged? The latter is what I want to preclude.

@Elliria At first, your idea of putting beta between main and develop did∕n't make any sense, but I think if we rename it from "beta" to "next" or something similar that does not directly imply that it's a "live" beta release, then it makes perfect sense.
That would be one step better than what we have now (once it's all synchronized so that all the commits are one directional, not both ways). However, it sounds like this approach (like our current approach) may be way more trouble than it should be. But, IDK enough to know which of the other suggested methods or something else would be better for us.

Here in NY, we have ridiculously complex and specific zoning laws and building codes when you want to build a house and really high taxes. We designed a house when we really wanted a couple of small cabins and no matter what we did we couldn't get the cost down anywhere near what we wanted. The laws blocked us. Later, I looked at a piece  of land in West Virginia. The taxes sounded similar until I found out that the figure quoted was for a whole year and not for a month! As for zoning/code compliance, I was told "Well, you have to live in it." That was it!

The moral of this aside is that both extremes are a bit much to deal with and we should figure out the right mix/level of structure for our project. We have to maintain release control/consistency, but AutoKey is not a mission critical application, etc., so we don't need any rocket science or additional complexity to assure that nothing can ever go wrong. I definitely don't want to stay on "method 5" for which the recommendatiion was, "run the other way" if we don't have to.

In this vein, @Elliria and @sebastiansam55 - "you have to live in it" as developers and unless I really object to something, I'll merge (or agree with you to merge) anything you ask me to once I understand what it is at a very basic level. I'm not qualified to review most things line by line, etc., but I'll take a look whenever I can...

**Reply by Elliria** [reply]
I would think that most package maintainers would find [our **Releases** page with download links](https://github.com/autokey/autokey/releases/) or would look for [the tags in our repository](https://github.com/autokey/autokey/tags) to build from. We're also working on gradually improving our documentation, which will be helpful for them if they read it. Last, but not least, we could do outreach to any maintainers who we're aware of if there's a change.

**Reply by Elliria** [reply]
The **beta** name is actually fine as it is, since it's an industry standard and is universally understood to be a new, experimental version that's believed to be good, but requires further real-world testing to be sure.

Since only solid code goes into a beta, it stands to reason that, regardless of whether its code is  tagged as an official beta release or not, anyone could (and should be encouraged to) build the **beta** branch code at any time and give it a whirl to try out what's currently being worked on. Its name lets them know that.

Also, the **beta** branch gives the AutoKey team the ability to point users to it as a source for code that provides specific features or solves specific problems that haven't yet been implemented in current official releases. So, if a user is in a hurry for _[insert feature or solution to problem here]_, they can try the beta.

They could also live dangerously and build from the **develop** branch at any time (if building even works), but they'd be taking their chances with code that's not necessarily good yet and/or may be unfinished, and we definitely wouldn't want to encourage that.

**Comment by Elliria** [comment]
From what I've been able to figure out so far, this is a very basic summary of our branching/workflow/release strategy (corrections are welcome):
1. Do development in a fork on **develop** or on feature branches off of **develop**.
2. Apply hotfixes, if needed, to hotfix branches off of **master**.
3. When ready for a release
   1. Update **develop**:
      * Cherry-pick any fixes from **master** onto **develop**.
   3. Pick one to update **beta**:
      * Cherry-pick:
        1. Do a hard reset on **beta** (`git reset --hard master`) to ensure that it's an identical snapshot of **master**.
        2. Cherry-pick commits we want to include in the release from **develop** onto **beta**.
      * Merge:
        1. Do a hard reset on **beta** (`git reset --hard master`) to ensure that it's an identical snapshot of **master**.
        2. Merge **develop** into **beta**.
      * Reset:
        1. Do a hard reset on **beta** (`git reset --hard develop`) to ensure that it's an identical snapshot of **develop**. 
4. Tag **beta** as a release (it will be automatically built by a workflow).
5. Turn the tagged **beta** release loose on the public for testing and feedback.
6. Pick one:
   * If changes are necessary:
     1. Make adjustments to **develop** based on the testing and feedback.
     2. Return to step 3 above.
   * If changes are not necessary:
     1. Go to the next step.
7. Do a hard reset on **master** (`git reset --hard beta`) to ensure that it's an identical snapshot of **beta**.
8. Tag **master** as a release (it will be automatically built and published to [PyPi](https://pypi.org/search/?q=autokey) by a workflow).
9. Announce the new AutoKey release to our contacts at the various distributions.
10. Let **beta** go dormant until the next release.
11. Return to step 1.

**Reply by josephj11** [reply]
I was definitely partly responsible for creating this mess. I want to get rid of develop as soon as we have extracted everything useful from it. I don't understand git tags very well, but if they do what I think they do, we may be able to replace develop and beta with appropriate tags on master. Its sounds so much simpler.

**Comment by Elliria** [comment]
The branches have been that way for a long time. We did make some commits to the **master** branch that we shouldn't have, but live and learn and it will be cleaned up.

Git is pretty simple under the hood. It's a collection of moments. Every commit is a moment in time that's represented by its commit ID. If you know the ID, you can use it to visit the code in that moment in time. Tags are labels that you can put onto any commit to make it easy to find (instead of having to remember a commit ID). You can use a label to visit the code in that moment in time. That's it in a nutshell.

One of our workflows watches for tags with the `v*.*.*` format and reacts to them:
* When it's triggered by the presence of such a tag on the **beta** branch:
  * It builds AutoKey from the **beta** branch and uploads the build as a [pre-release (a beta release)](https://github.com/autokey/autokey/releases). 
* When it's triggered by the presence of such a tag on the **master** branch:
  * It builds AutoKey from the **master** branch and uploads the build as a [release (an official release)](https://github.com/autokey/autokey/releases).
  * It publishes the build to the [Python Package Index](https://pypi.org/).

I'm not sure what, if anything, happens if such a tag were used on the **develop** branch, but I think it just wouldn't be done.

**Reply by josephj11** [reply]
That makes it very clear. Thanks. How do merges fit into this explanation? Are they just "importing" commits from another branch?

**Comment by Elliria** [comment]
In **git** (locally on your own computer):
* If the parent branch and the feature branch have each changed/diverged, then the merge **is** a commit. That's known as a **true merge**.
* If only the feature branch has changed/diverged, then the merge **is not** a commit. That's known as a **fast-forward merge**.

On **GitHub**, that's simplified or complicated (depending on perspective) a bit, because I'm pretty sure that every  merge is a commit regardless of circumstance.

**Comment by BlueDrink9** [comment]
> From what I've been able to figure out so far, this is a very basic summary of our branching/workflow/release strategy (corrections are welcome):
> 1. Do development in a fork on **develop** or on feature branches off of **develop**.
> 2. Apply hotfixes, if needed, to hotfix branches off of **master**.
> 3. When ready for a release
>    1. Update **develop**:
>       * Cherry-pick any fixes from **master** onto **develop**.
>    3. Pick one to update **beta**:
>       * Cherry-pick:
>         1. Do a hard reset on **beta** (`git reset --hard master`) to ensure that it's an identical snapshot of **master**.
>         2. Cherry-pick commits we want to include in the release from **develop** onto **beta**.
>       * Merge:
>         1. Do a hard reset on **beta** (`git reset --hard master`) to ensure that it's an identical snapshot of **master**.
>         2. Merge **develop** into **beta**.
>       * Reset:
>         1. Do a hard reset on **beta** (`git reset --hard develop`) to ensure that it's an identical snapshot of **develop**. 
> 4. Tag **beta** as a release (it will be automatically built by a workflow).
> 5. Turn the tagged **beta** release loose on the public for testing and feedback.
> 6. Pick one:
>    * If changes are necessary:
>      1. Make adjustments to **develop** based on the testing and feedback.
>      2. Return to step 3 above.
>    * If changes are not necessary:
>      1. Go to the next step.
> 7. Do a hard reset on **master** (`git reset --hard beta`) to ensure that it's an identical snapshot of **beta**.
> 8. Tag **master** as a release (it will be automatically built and published to [PyPi](https://pypi.org/search/?q=autokey) by a workflow).
> 9. Announce the new AutoKey release to our contacts at the various distributions.
> 10. Let **beta** go dormant until the next release.
> 11. Return to step 1.
> 

This is an *excellent* overview Eliria, and matches basically exactly what I set up beta for. It served its purpose well for getting the massive overhauls of 0.96 over the line, but you could definitely delete it if you wanted to and don't anticipate a need for pre-releases.

**Comment by BlueDrink9** [comment]
> In **git** (locally on your own computer):
> * If the parent branch and the feature branch have each changed/diverged, then the merge **is** a commit. That's known as a **true merge**.
> * If only the feature branch has changed/diverged, then the merge **is not** a commit. That's known as a **fast-forward merge**.
> 
> On **GitHub**, that's simplified or complicated (depending on perspective) a bit, because I'm pretty sure that every  merge is a commit regardless of circumstance.
> 

Fantastic. Tiny correction: not always a fast-forward merge, you can explicitly request with `--no-fast-forward` that a non-diverged merge still be made without fast forwarding (i.e., as a regular merge commit).

There's also rebasing, which is closer to a fast forward but for diverged branches. A bit more complicated, probably don't need to pay attention to it
